File: http/http_golang/timer.go
package main

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"strings"
	"time"
)

type BiliTimer struct {
	SaleTime  float64
	DelayTime float64
	Message   []byte
	client    *tls.Conn
}

func (receiver *BiliTimer) init(SaleTime, DelayTime int64) *BiliTimer {
	receiver.DelayTime = float64(DelayTime / 1000)
	receiver.SaleTime = float64(SaleTime)
	var MessageList = []string{
		"GET /x/report/click/now HTTP/1.1\r\nhost: api.bilibili.com", "Connection: keep-alive",
		"User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0",
	}
	receiver.Message = []byte(strings.Join(MessageList, "\r\n") + "\r\n\r\n")
	return receiver
}

func (receiver *BiliTimer) updateClient() {
	receiver.client, _ = tls.Dial("tcp", "api.bilibili.com:443", &tls.Config{
		InsecureSkipVerify: true,
		ServerName:         "api.bilibili.com",
		MinVersion:         tls.VersionTLS12,
		MaxVersion:         tls.VersionTLS12,
		ClientAuth:         tls.RequireAndVerifyClientCert,
	})
}

func (receiver *BiliTimer) GetBiliTime() float64 {
	_, _ = receiver.client.Write(receiver.Message)
	var buf = make([]byte, 1024)
	var length, _ = receiver.client.Read(buf)
	var rec = string(buf[:length])
	if len(rec) == 0 {
		receiver.updateClient()
		return receiver.GetBiliTime()
	}
	var JsonData = make(map[string]map[string]int64)
	var SplitBody = strings.Split(rec, "\r\n\r\n")
	var Body = SplitBody[len(SplitBody)-1]
	var _ = json.Unmarshal([]byte(Body), &JsonData)
	return float64(JsonData["data"]["now"])
}

func (receiver *BiliTimer) WaitLocalTime(jump int64) float64 {
	var NowTime, JumpTime float64
	NowTime = float64(time.Now().UnixNano()) / 1e9
	JumpTime = receiver.SaleTime - float64(jump)
	for JumpTime > NowTime {
		fmt.Printf("\r%f", NowTime)
		NowTime = float64(time.Now().UnixNano()) / 1e9
	}
	return NowTime
}

func (receiver *BiliTimer) WaitSeverTime() float64 {
	var SleepTimeNumber = receiver.DelayTime * float64(time.Second)
	receiver.updateClient()
	var NowTime = receiver.GetBiliTime()
	for NowTime < receiver.SaleTime {
		fmt.Printf("\r%f", NowTime)
		NowTime = receiver.GetBiliTime()
	}
	time.Sleep(time.Duration(SleepTimeNumber))
	return NowTime
}

File: http/source/python/timer.py
import socket
import json
import time
import ssl

host = "api.bilibili.com"

message_list = [
    f"GET /x/report/click/now HTTP/1.1", f"host: {host}", "Connection: keep-alive",
    "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0",
]

http_message = str("\r\n".join(message_list) + "\r\n\r\n").encode()


class BiliTimer(object):
    def __init__(self, sale_time: int, delay_time: int):
        super(BiliTimer, self).__init__()
        self.sale_time = float(sale_time)
        self.delay_time = delay_time / 1000
        self.client = None

    def updateClient(self, **kwargs):
        context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
        context.purpose = ssl.Purpose.SERVER_AUTH
        context.verify_mode = ssl.CERT_REQUIRED
        context.check_hostname = True
        context.load_default_certs()
        connection = socket.create_connection((host, 443))
        kwargs.update({"server_hostname": host})
        self.client = context.wrap_socket(connection, **kwargs)

    def GetBiliTime(self):
        self.client.sendall(http_message)
        rec = self.client.read()
        if rec == b"":
            self.updateClient()
            return self.GetBiliTime()
        body = rec.split(b"\r\n\r\n")[-1]
        body_json = json.loads(body.decode())
        return int(body_json["data"]["now"])

    def WaitLocalTime(self, jump_time: int):
        now_time = time.time()
        jump_to_time = self.sale_time - jump_time
        while jump_to_time > now_time:
            print(f"\r{now_time}", end="")
            now_time = time.time()
            time.sleep(0.001)
        return jump_to_time

    def WaitSeverTime(self):
        self.updateClient()
        now_time = self.GetBiliTime()
        while self.sale_time > now_time:
            print(f"\r{now_time}", end="")
            now_time = self.GetBiliTime()
        time.sleep(self.delay_time)
        return now_time
